Optional function in generic structure cannot be compiled

### Describe the bug

If a generic structure has an optional function with generic type. It cannot be compiled by the latest v compiler, but can be compiled by v 0.4.

### Reproduction Steps

Compile this file.

```v
module main

struct Cmp[K] {
	cmp ?fn (K, K) bool
}

fn (c Cmp[K]) compare(a K, b K) bool {
	cmp := c.cmp or { return a < b }
	return cmp(a, b)
}

fn main() {
	c := Cmp[int]{}
	print(c.compare(1, 2))
}

```

### Expected Behavior

print true

### Current Behavior

Compile error

```
==================
/tmp/v_501/test_v.13496298430663525765.tmp.c:12438:28: error: initializing '_option_anon_fn_k_k__bool' (aka 'struct _option_anon_fn_k_k__bool') with an expression of incompatible type '_option_anon_fn_int_int__bool' (aka 'struct _option_anon_fn_int_int__bool')
        _option_anon_fn_k_k__bool _t1 = c.cmp;
                                  ^     ~~~~~
1 error generated.

```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.0 5e0ebba

### Environment details (OS name and version, etc.)

MacOS with M1 chip

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3603,7 +3603,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	if node.or_block.kind != .absent && !g.is_assign_lhs && g.table.sym(node.typ).kind != .chan {
 		is_ptr := sym.kind in [.interface_, .sum_type]
 		stmt_str := g.go_before_stmt(0).trim_space()
-		styp := g.typ(node.typ)
+		styp := g.typ(g.unwrap_generic(node.typ))
 		g.empty_line = true
 		tmp_var := g.new_tmp_var()
 		g.write('${styp} ${tmp_var} = ')

--- a/vlib/v/tests/generics_struct_with_option_fn_test.v
+++ b/vlib/v/tests/generics_struct_with_option_fn_test.v
@@ -1,0 +1,14 @@
+struct Cmp[K] {
+	cmp ?fn (K, K) bool
+}
+
+fn (c Cmp[K]) compare(a K, b K) bool {
+	cmp := c.cmp or { return a < b }
+	return cmp(a, b)
+}
+
+fn test_generic_struct_with_option_fn() {
+	c := Cmp[int]{}
+	print(c.compare(1, 2))
+	assert c.compare(1, 2)
+}

